cpp/src/arrow/scalar_test.cc

```diff
@@ -1106,6 +1106,25 @@ class TestListScalar : public ::testing::Test {
     ASSERT_RAISES(Invalid, scalar.ValidateFull());
   }
 
+  void TestHashing() {
+    ScalarType empty_bitmap_scalar(ArrayFromJSON(int16(), "[1, 2, 3]"));
+    ASSERT_OK(empty_bitmap_scalar.ValidateFull());
+    ASSERT_TRUE(empty_bitmap_scalar.value->data()->buffers[0] == nullptr);
+    
+    auto data = empty_bitmap_scalar.value->data()->buffers[1];
+    std::vector<uint8_t> bitmap_data = {0,0,0};
+    auto null_bitmap = std::make_shared<Buffer>(bitmap_data.data(), 3);
+
+    std::shared_ptr<Int16Array> arr(new Int16Array(3, data, null_bitmap, 0));
+    ASSERT_TRUE(arr->null_count() == 0);
+    // this line fails - I don't know how to create an array with a null bitmap
+    // that is all 0s.
+    ASSERT_TRUE(arr->data()->buffers[0] != nullptr);
+    ScalarType set_bitmap_scalar(arr);
```

Comment:
As I said in the description, I don't know how to make an array with a 0s null bitmap. I ran into this bug when using the `filter()` function on a `LargeListArray` of strings. The scalars in the array would have the null bitmap set even if there were no nulls, changing the hash values. I assume using that exact method is too circumspect to be used here.

Reply:
Yeah, it seems we explicitly don't keep the bitmap around if there are no nulls in the Array construction:

https://github.com/apache/arrow/blob/ec29c6ffc3cb1af4db4903d9877b2f0b548a3ad9/cpp/src/arrow/array/data.cc#L53-L55

One way around this would be to create a ListArray, and then get one element of it as a scalar (like my code snippet in python)

Reply:
Yes, just rewrite the Python snippet (or some equivalent) in C++.  It should actually be easy, using `ArrayFromJSON` and `Array::GetScalar`.